src/skelet_cps.rs

```diff
@@ -0,0 +1,376 @@
+// reimplementation of reimplementation of skelet's implementation :) https://gist.github.com/savask/1c43a0e5cdd81229f236dcf2b0611c3f
+
+use anyhow::{Context, Result};
+use hashbrown::HashMap;
+use indexmap::IndexSet;
+use std::{fmt, str};
+
+use crate::{
+    interner::{ITape, InternedDisplay, InternerTape},
+    machine::{Head, Machine, Orientation},
+    ui_dbg, ProverResult,
+};
+
+#[derive(Debug)]
+pub enum Err {
+    Halt,
+    Loop,
+    ConfLimit,
+}
+
+type Interner = InternerTape<u8>;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq, Hash)]
+struct Configuration {
+    /// 0 == left, 1 == right, 2 == middle
+    tape: [ITape; 3],
+    head: Head,
+}
+
+impl Configuration {
+    pub fn new(zeros: ITape) -> Configuration {
+        // 0..0 0..0 A> 0..0
+        Configuration { tape: [zeros, zeros, zeros], head: Head { state: 0, orient: 1 } }
+    }
+
+    /// runs until head leaves the tape, split & return new conf, C (`<S ABC` -> `<S AB` & `C`) & end conf string if in tui mode
+    fn run(
+        mut self,
+        machine: &Machine,
+        mut step_limit: usize,
+        segment_size: usize,
+        interner: &mut Interner,
+    ) -> Result<(Configuration, ITape, String), Err> {
+        // TODO(perf): switch to single tape?
+        let mut tape = [interner[self.tape[0]].to_vec(), interner[self.tape[1]].to_vec()];
+        tape[self.head.op_orient()].extend_from_slice(&interner[self.tape[2]]);
+        fn format_conf(tape: &[Vec<u8>; 2], head: Head) -> String {
+            format!(
+                "{}{}{}",
+                tape[0].iter().map(u8::to_string).collect::<String>(),
+                head,
+                tape[1].iter().rev().map(u8::to_string).collect::<String>(),
+            )
+        }
+        while let Some(symbol) = tape[self.head.orient as usize].pop() {
+            let trans = machine.get_transition(symbol, self.head.state).ok_or(Err::Halt)?;
+            if trans.head.state >= machine.states() {
+                return Err(Err::Halt);
+            }
+            self.head = trans.head;
+            tape[1 - self.head.orient as usize].push(trans.symbol);
+            ui_dbg!("\t\t{}", format_conf(&tape, self.head));
+            step_limit -= 1;
+            if step_limit == 0 {
+                return Err(Err::Loop);
+            }
+        }
+        let s = if cfg!(not(feature = "ui_tui")) { String::new() } else { format_conf(&tape, self.head) };
+        // `[] <S AB C`, last item on tape is closest to head
```

Comment:
yes, comment is confusing, because it's all confusing :). tape is represented as 2 stacks, but i use vector (== list) as stack impl. `push()` and `pop()` works as expected (top (first) stack element == last vector item), but here i use indexing ([`.chunks().rev()`](https://github.com/univerz/bbc/pull/2/files/325f5d596e9024411278571b7a1bc6721d0c3fed#diff-40c1ea6df801ba5f3e88bf2be1f82b45e9cb741634a3e625a286ef7339833087R71)) and the problem with confusing first & last item is visible.